scripts/build/termux_step_massage.sh

```diff
@@ -82,7 +82,7 @@ termux_step_massage() {
 			if head -c 100 "$file" | head -n 1 | grep -E "^#!.*/bin/.*" | grep -q -E -v -e "^#! ?/system" -e "^#! ?$TERMUX_PREFIX_CLASSICAL"; then
 				sed --follow-symlinks -i -E "1 s@^#\!(.*)/bin/(.*)@#\!$TERMUX_PREFIX/bin/\2@" "$file"
 			fi
-		done < <(find -L . -type f -print0)
+		done < <(grep --recursive --binary-files=without-match --files-with-matches --null --max-count=1 '^#' .)
```

Comment:
I think there should still be a comment here.
Doesn't have to too anything too long or detailed.
Just something along the lines of:
```suggestion
		# Match all text files containing a `#` at the start of any line.
		done < <(grep --recursive --binary-files=without-match --files-with-matches --null --max-count=1 '^#' .)
```

Reply:
Can you please suggest a comment for the second place too? My comments are awful.

Reply:
> `at the start of any line.`

Actually no. At the start of the first line. That is the point.

Reply:
Wait a sec, maybe I am wrong.

Reply:
Your `grep` command as written searches the entire file, stopping after 1 match.

Reply:
Oh, ok. I hoped it searches searches only in the first line but I was wrong. 
https://github.com/termux/termux-packages/pull/24518#pullrequestreview-2797718200 thunder-coding thinks it is good without additional comments. Who should I agree with?

Reply:
> Who should I agree with?

I can't make that decision for you.
But I can make the case for comments.
https://github.com/termux/termux-packages/pull/24518#issuecomment-2833597493

Reply:
I think I will agree with you. Can you please suggest a comment for the second place too? My comments are awful.

Reply:
> My comments are awful.

I wouldn't say your comments are bad.
They tend to be verbose and matter of fact.
That isn't bad, but it can make them harder to understand.

Writing comments is an exercise in communication and it's just as important a skill to exercise as any other part of development.

Reply:
Seems like I found solution to look for matches at the start of file, but it is awful.
```bash
# Perl regexp allowing searching at the start of file
$ time grep --recursive --perl-regexp --binary-files=text --files-with-matches --max-count=1 '\A\x7fELF' /data &> /dev/null
real    0m0.780s
user    0m0.687s
sys     0m0.093s

# Current solution
$ time grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' /data &> /dev/null
real    0m0.156s
user    0m0.040s
sys     0m0.116s
```

But there is ugrep which is a bit faster
```bash
$ time ugrep --recursive --binary-files=text --files-with-matches --max-count=1 --fixed-strings $'\x7fELF' /data &> /dev/null
real    0m0.077s
user    0m0.102s
sys     0m0.207s

$ time ugrep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' /data &> /dev/null
real    0m0.100s
user    0m0.130s
sys     0m0.282s
```

Unfortunately it is not installed and I am pretty much sure we should not install it only for massaging step in the case if we win only 50-70ms here.
But in the case if it is a full drop-in replacement to regular grep we may consider replacing/linking (with regular `grep() { ugrep "$@"; }`) to speed-up everything a little bit.

Reply:
It needs to be `--binary-files=text`, binary files can also contain a shebang, like python executable files like `youtube-dl`. `yt-dlp`. etc. 

- http://blog.ablepear.com/2012/10/bundling-python-files-into-stand-alone.html

Also use `'^#!.'`, only matching a `#` would match files with a comment on first line. The dot is to match at least one character that should be replaced.

Reply:
As I said before in different review thread, it does not replace actual filtering, but it performs some very basic prefiltering to avoid binary execution overhead in actual filtering.

Reply:
Should not be shebang be relevant only to text files?

Reply:
And if you are gonna spawn a `grep` process to check for shebang, might as well do it strictly/correctly.

Reply:
> Should not be shebang be relevant only to text files?

`yt-dlp` is possibly not a good example (because in Termux i cannot see where its shebang-containing binary files are, there might not be any for that example), but agnostic-apollo is right, it is possible that a file which `grep --binary-files=without-match` excludes might contain an instance of `#!/usr` which should be replaced. I think I have seen one somewhere, once, a very long time ago, but I can't remember what software it was in exactly.

Reply:
Additionally if you really wanted to save time, do following where only one `grep` is run, instead of running first `grep` for prefilter, second `grep` for extracting shebang, third `grep` for excluding already replaced shebang files, and another `grep` for if `$TERMUX_PKG_NO_SHEBANG_FIX_FILES` is set. In following only `grep` is run and other regexes are done in shell itself, with proper escaping. Additionally, `$TERMUX_PKG_NO_SHEBANG_FIX_FILES` can be a list of regex instead of individual files.

`qt6-qbase` package should escape dot `.` in `$TERMUX_PKG_NO_SHEBANG_FIX_FILES`.

```shell
if [ "$TERMUX_PKG_NO_SHEBANG_FIX" != "true" ]; then
	local no_shebang_replace_full_regex="" no_shebang_replace_regex
	# Assume files are a regex list and special characters in paths in each regex are already escaped.
	if [[ -n "$TERMUX_PKG_NO_SHEBANG_FIX_FILES" ]]; then
		while IFS= read -r no_shebang_replace_regex; do
	        [[ -n "$no_shebang_replace_full_regex" ]] && no_shebang_replace_full_regex+='|'
			no_shebang_replace_full_regex+="($no_shebang_replace_regex)"
	    done < <(printf "%s\n" "$TERMUX_PKG_NO_SHEBANG_FIX_FILES")
		if [[ -n "$no_shebang_replace_full_regex" ]]; then
			no_shebang_replace_full_regex='(|./)('"$no_shebang_replace_full_regex"')$'
		fi
	fi

	# Escape '\$[](){}|^.?+*' with backslashes.
	local prefix_escaped shebang_already_valid_regex shebang_match

	prefix_escaped="$(printf "%s" "$TERMUX_PREFIX_CLASSICAL" | sed -zE -e 's/[][\.|$(){}?+*^]/\\&/g')"
	shebang_already_valid_regex='^#! ?((/system/)|('"$prefix_escaped"'/))'

	# Fix shebang paths:
	while IFS= read -r -d '' file; do
		shebang_match="$(grep --binary-files=text --null --max-count=1 '^#!.*/bin/.*' "$file" || true)"
		if [[ -n "$shebang_match" ]]; then
			if [[ "$shebang_match" =~ $shebang_already_valid_regex ]]; then
				echo "INFO: Skip shebang fix for '$file' as shebang '$shebang_match' already valid"
				continue
			fi
			if [[ -n "$no_shebang_replace_full_regex" ]] && [[ "$file" =~ $no_shebang_replace_full_regex ]]; then
				echo "INFO: Skip shebang fix for '$file' as its excluded"
				continue
			fi
			sed --follow-symlinks -i -E "1 s@^#\!(.*)/bin/(.*)@#\!$TERMUX_PREFIX/bin/\2@" "$file"
		fi
	done < <(find -L . -type f -print0)
fi
```

Reply:
And for reference, https://github.com/stargateoss/temporal-src-network/blob/c7b9e24871a2e1f98af7fd32ae667046b0afcdb4/site/pages/en/docs/build/index.md details 3 ways of executing python multi-source-file project from a single file.

Reply:
@agnostic-apollo Emmmm... Your suggestion completely eliminates the optimization and even makes the thing worse since it spawns a lot more processes.

The whole point of the optimization is to spawn a single relatively lightweight CPU-cheap command which will filter-out obviously non-matching results and prevent iterating over all possible files.

But seems like I found better design for preliminary check.
```
find . -type f -print0 | perl -0ne 'open F,"<",$_;read(F,$b,2);print "$_\0" if $b eq "#!";'
```
in the place of `find . -type f -print0` effectively filters all possible files starting with `#!` symbols even though it is hardly readable.

Also we can use similar commands for searching ELFs and static archives.
```
find . -type f -print0 | perl -0ne 'open F,"<",$_;read(F,$b,4);print "$_\0" if $b eq "\x7fELF" or $b eq "!<ar";'
```

Instead of scanning the whole file for expression match (like in `grep`-based solution) `perl` reads 2 (for shebang) or 4 (for elfs and static archives) bytes and compares them with needed signatures.

---

This code
```
TERMUX_PREFIX_CLASSICAL=/data/data/com.termux/files/usr
TERMUX_PREFIX=/data/data/com.termux/files/usr
cd $TERMUX_PREFIX
while IFS= read -r -d '' file; do
    head -c 100 "$file" | head -n 1 | grep -E "^#!.*/bin/.*" | grep -q -E -v -e "^#! ?/system" -e "^#! ?$TERMUX_PREFIX_CLASSICAL" && echo "${file}"
done < <(find -L . -type f -print0)
```
takes a lot of time in termux docker with all possible packages installed:
```
real    23m40.131s
user    28m23.800s
sys     41m33.801s
```

But the script's execution time drops dramatically when we add the preliminary check.
```
real    0m15.040s
user    0m18.140s
sys     0m27.068s
```

The filtering itself works mind-blowingly fast and has similar results for filtering elfs/static archives and shebang-starting files.
```
$ time find /data -type f -print0 | perl -0ne 'open F,"<",$_;read(F,$b,2);print "$_\0" if $b eq "#!";' &> /dev/null

real    0m1.366s
user    0m0.423s
sys     0m1.266s
```

I did not check it with `kf6-breeze-icons` yet, only tested with termux docker on PC. But speed of scanning 12GB of files impressive.

Reply:
>Emmmm... Your suggestion completely eliminates the optimization and even makes the thing worse since it spawns a lot more processes.

Sorry, you are right, it would always spawn one `grep` process for each file instead of filtering out all files with one `grep`.

However, you would be adding a `perl` dependency to our internal scripts, it does not seem to be used anywhere else, although its listed in `setup-termux.sh`, possibly for `build.sh` files.

Instead use following where file header is read in `bash` itself with `read` builtin and matched with builtin regex. `old-shebang` is for `grep` solution I posted before, and there is a *slight* improvement from before. This should be more than sufficient for a small massage directory in terms of efficiency.

```shell
local file

if [ "$TERMUX_PKG_NO_SHEBANG_FIX" != "true" ]; then
	local no_shebang_replace_full_regex="" no_shebang_replace_regex
	# Assume files are a regex list and special characters in paths in each regex are already escaped.
	if [[ -n "$TERMUX_PKG_NO_SHEBANG_FIX_FILES" ]]; then
		while IFS= read -r no_shebang_replace_regex; do
	        [[ -n "$no_shebang_replace_full_regex" ]] && no_shebang_replace_full_regex+='|'
			no_shebang_replace_full_regex+="($no_shebang_replace_regex)"
	    done < <(printf "%s\n" "$TERMUX_PKG_NO_SHEBANG_FIX_FILES")
		if [[ -n "$no_shebang_replace_full_regex" ]]; then
			no_shebang_replace_full_regex='(|./)('"$no_shebang_replace_full_regex"')$'
		fi
	fi

	local prefix_escaped shebang_already_valid_regex header_line shebang_match
	local shebang_regex='^#!.*/bin/.*'

	# Escape '\$[](){}|^.?+*' with backslashes.
	prefix_escaped="$(printf "%s" "$TERMUX_PREFIX_CLASSICAL" | sed -zE -e 's/[][\.|$(){}?+*^]/\\&/g')"
	shebang_already_valid_regex='^#! ?((/system/)|('"$prefix_escaped"'/))'

	# Fix shebang paths:
	while IFS= read -r -d '' file; do
		read -r header_line < "$file"
		if [[ "${header_line:0:1}" == "#" ]] && [[ "${header_line:1:1}" == "!" ]] && \
			[[ "${#header_line}" -ge 3 ]] && \
			[[ "$header_line" =~ $shebang_regex ]]; then
			shebang_match="${BASH_REMATCH[0]}"
			if [[ -n "$shebang_match" ]]; then
				if [[ "$shebang_match" =~ $shebang_already_valid_regex ]]; then
					echo "INFO: Skip shebang fix for '$file' as shebang '$shebang_match' already valid"
					continue
				fi
				if [[ -n "$no_shebang_replace_full_regex" ]] && [[ "$file" =~ $no_shebang_replace_full_regex ]]; then
					echo "INFO: Skip shebang fix for '$file' as its excluded"
					continue
				fi
				sed --follow-symlinks -i -E "1 s@^#\!(.*)/bin/(.*)@#\!$TERMUX_PREFIX/bin/\2@" "$file"
			fi
		fi
	done < <(find -L . -type f -print0)
fi
```

```shell
$ du -sh /usr/lib
18G	/usr/lib

$ find -L /usr/lib -type f 2>/dev/null | wc -l
773452

$ time bash ./old-shebang &>/dev/null
real    30m37.331s
user    21m55.018s
sys     8m1.811s

$ time bash ./new-shebang &>/dev/null
real    0m44.607s
user    0m12.417s
sys     0m20.172s
```


For ELF and static library matching, do following.

```shell
local file
local header_line
local -a files

while IFS= read -r -d '' file; do
	read -r header_line < "$file"
	if [[ "$header_line" == $'\x7fELF'* ]] || [[ "$header_line" == '!<arch>'* ]]; then
		files+=("$file")
	fi
done < <(find -L . -type f -print0)
```

```shell
$ time bash ./elf &>/dev/null

real    0m32.559s
user    0m10.925s
sys     0m17.672s

$ bash ./elf 2>/dev/null | wc -l
125650
```

Reply:
Oh, it is wonderful! I was afraid bash paring overhead will be higher than in perl.
But I'd like to NOT use `-L` because in some cases symlinks are invalid in massagedir since `tar` copies symlinks without following them.

Also it seems like I found a slightly better solution, a little bit slower than perl solution.
```
$ sudo sync; sudo sysctl -w vm.drop_caches=3; time find /usr/lib -type f -print0 | perl -0ne 's/\0$//;open F,"<",$_;read(F,$b,4);print "$_\n" if $b eq "\x7fELF" or $b eq "!<ar";' &> /dev/null
vm.drop_caches = 3
real    0m10.230s
user    0m0.559s
sys     0m2.083s

$ sudo sync; sudo sysctl -w vm.drop_caches=3; time (IFS=; find /usr/lib -type f -print0 | while read -r -d '' f; do read -r hdr <"$f"; [[ $hdr == $'\x7fELF'* || $hdr == '!<ar'* ]] && printf '%s\0' "$f"; done &> /dev/null )
vm.drop_caches = 3
real    0m23.136s
user    0m4.861s
sys     0m5.693s

$ sudo sync; sudo sysctl -w vm.drop_caches=3; time (IFS=; find /usr/lib -type f -print0 | while read -r -d '' f; do read -rN4 hdr <"$f"; [[ $hdr == $'\x7fELF' || $hdr == '!<ar' ]] && printf '%s\0' "$f"; done &> /dev/null )
vm.drop_caches = 3

real    0m13.927s
user    0m2.256s
sys     0m5.270s

```

Reply:
> But for shebang, please use the code I posted as it fixes escaping prefix and adds support for no shebang file list regex, and also logs properly, and removes external calls to grep/head. Just add N4, and remove -L if needed.

Ok, but I changed 
```
...
while IFS= read -r -d '' file; do
	read -r header_line < "$file"
	if [[ "${header_line:0:1}" == "#" ]] && [[ "${header_line:1:1}" == "!" ]] && \
		[[ "${#header_line}" -ge 3 ]] && \
		[[ "$header_line" =~ $shebang_regex ]]; then
...
```
to 
```
...
while IFS= read -rd '' file; do
	read -rn100 header_line < "$file"
	if [[ "${header_line:0:2}" == "#!" && "${#header_line}" -ge 3 && "$header_line" =~ $shebang_regex ]]; then
...
```
to make it a bit shorter and to limit size of line we read to 100 first symbols.
This way this part matches old behaviour and becomes more efficient on large files without newlines.

Reply:
> For ELF, you can use a shorter form.

I am not sure what exactly you mean, shorter form will be less readable.

Reply:
worth mentioning at least briefly that if `perl` is ever removed from the ubuntu docker image, it would remove a huge amount of things including important packages like `autoconf`, `git`.

```
The following packages will be REMOVED:
  apache2-bin apt-file autoconf autogen automake
  build-essential docbook docbook-dsssl docbook-to-man
  docbook-utils docbook-xml dpkg-dev git gnome-common
  gobject-introspection gobject-introspection-bin
  gtk-doc-tools guile-3.0-dev help2man intltool
  libapache2-mod-php8.3 libapt-pkg-perl libclone-perl
  libdpkg-perl libencode-locale-perl liberror-perl
  libfile-homedir-perl libfile-listing-perl
  libfile-which-perl libgirepository-1.0-dev
  libgirepository1.0-dev libhtml-parser-perl
  libhtml-tagset-perl libhtml-tree-perl
  libhttp-cookies-perl libhttp-date-perl
  libhttp-message-perl libhttp-negotiate-perl
  libio-html-perl libio-socket-ssl-perl libjson-perl
  libltdl-dev liblwp-mediatypes-perl
  liblwp-protocol-https-perl libmime-charset-perl
  libnet-http-perl libnet-ssleay-perl
  libparse-yapp-perl libpod-parser-perl
  libregexp-assemble-perl librsvg2-dev libsgmls-perl
  libsyntax-keyword-try-perl libtext-unidecode-perl
  libtimedate-perl libtry-tiny-perl
  libunicode-linebreak-perl liburi-perl libwww-perl
  libwww-robotrules-perl libxml-libxml-perl
  libxml-namespacesupport-perl libxml-parser-perl
  libxml-sax-base-perl libxml-sax-perl
  libxs-parse-keyword-perl libyaml-tiny-perl perl po4a
  python3-myst-parser python3-recommonmark
  python3-sphinx python3-sphinx-rtd-theme
  python3-sphinxcontrib.jquery
  python3-sphinxcontrib.qthelp sgml-data sgmlspl
  sphinx-common texinfo texinfo-lib texlive-base
  texlive-binaries texlive-extra-utils
  texlive-fonts-recommended texlive-formats-extra
  texlive-latex-base texlive-latex-extra
  texlive-latex-recommended texlive-luatex
  texlive-metapost texlive-pictures
  texlive-plain-generic texlive-xetex tipa triehash
  xmlto xmltoman
```

This also applies to the installation of Gentoo I am using, where removing `perl` would require removing over 300 packages.

`perl` can most likely only be removed from esoteric, very minimalist GNU/Linux distros, if ever.